crates/compiler/types/src/types.rs

```diff
@@ -3281,6 +2925,208 @@ impl Type {
     }
 }
 
+struct InstantiateAliasesCtx<'a> {
+    var_store: &'a mut VarStore,
+    new_lambda_set_variables: &'a mut ImSet<Variable>,
+    new_recursion_variables: &'a mut ImSet<Variable>,
+    new_infer_ext_vars: &'a mut ImSet<Variable>,
+}
+
+fn instantiate_aliases<'a, F>(
+    typ: &mut Type,
+    region: Region,
+    aliases: &F,
+    ctx: &mut InstantiateAliasesCtx<'_>,
+) where
+    F: Fn(Symbol) -> Option<&'a Alias>,
+{
+    use Type::*;
+
+    match typ {
+        Function(args, closure, ret) => {
+            for arg in args {
+                instantiate_aliases(arg, region, aliases, ctx);
+            }
+            instantiate_aliases(closure, region, aliases, ctx);
+            instantiate_aliases(ret, region, aliases, ctx);
```

Comment:
it looks like this could be written using an explicit stack, would that not be better? it depends a bit on when this is used (e.g. user-defined types are likely quite small, but internal ones can grow pretty big)

Reply:
yes, that's a good idea, let's do it in a follow-up.